### ISSUE TITLE
Segmentation fault error

### ISSUE THREAD

**Issue by patofet** [gautamdhameja/substrate-validator-set#22]
We found a bug using this pallet that sometimes when we add a node the node that I'm connected stops abruptly with the message "Segmentation fault" and then I must start it, then works fine with the new node added. I don't have the exact log of the console the next time I'll take it but, I remember that only shows "Segmentation fault" or something like that.

Edit: Now I was trying to remove a validator and give me this error on all validator nodes:
```
2021-10-07 09:58:22 ✨ Imported #8914 (0x2f76…b864)
2021-10-07 09:58:24 💤 Idle (3 peers), best: #8914 (0x2f76…b864), finalized #6991 (0x059e…bc78), ⬇ 0.6kiB/s ⬆ 0.5kiB/s
2021-10-07 09:58:29 💤 Idle (3 peers), best: #8914 (0x2f76…b864), finalized #6991 (0x059e…bc78), ⬇ 60 B/s ⬆ 56 B/s
2021-10-07 09:58:30 🙌 Starting consensus session on top of parent 0x2f76c381fb54c0c288ed5df5ab36ed6060fc0c6031e005f02e45182a771eb864
2021-10-07 09:58:30 🎁 Prepared block for proposing at 8915 [hash: 0x95c07ae37b787ab31a990dc43643b3253f3987ec04b300c2031e55c518441463; parent_hash: 0x2f76…b864; extrinsics (2): [0x6666…39b1, 0x3765…dc03]]
2021-10-07 09:58:30 🔖 Pre-sealed block for proposal at 8915. Hash now 0xa89ea3b6fc2df381c8106bbd0203db53e5f4415b9c151dcad63ab8c4633aa70e, previously 0x95c07ae37b787ab31a990dc43643b3253f3987ec04b300c2031e55c518441463.
2021-10-07 09:58:30 ✨ Imported #8915 (0xa89e…a70e)
2021-10-07 09:58:34 💤 Idle (3 peers), best: #8915 (0xa89e…a70e), finalized #6991 (0x059e…bc78), ⬇ 0.4kiB/s ⬆ 0.7kiB/s
2021-10-07 09:58:36 ✨ Imported #8916 (0xa670…2511)
2021-10-07 09:58:39 💤 Idle (3 peers), best: #8916 (0xa670…2511), finalized #6991 (0x059e…bc78), ⬇ 0.3kiB/s ⬆ 0.3kiB/s
2021-10-07 09:58:42 ✨ Imported #8917 (0x03fd…8479)
2021-10-07 09:58:44 💤 Idle (3 peers), best: #8917 (0x03fd…8479), finalized #6991 (0x059e…bc78), ⬇ 0.6kiB/s ⬆ 0.4kiB/s
2021-10-07 09:58:46 ✨ Imported #8918 (0xfd07…940c)
2021-10-07 09:58:49 💤 Idle (3 peers), best: #8918 (0xfd07…940c), finalized #6991 (0x059e…bc78), ⬇ 0.2kiB/s ⬆ 0.2kiB/s
2021-10-07 09:58:54 🙌 Starting consensus session on top of parent 0xfd077c119e7f86bac6885f7b4d280e6567064fdfeb79bcc2528c865249bd940c
2021-10-07 09:58:54 panicked at 'swap_remove index (is 3) should be < len (is 3)', library/alloc/src/vec/mod.rs:1297:13
Segmentation fault
```
After this error, every time that I start the nodes it give me this error, after this, I must to delete the two validators nodes and syncronize with the non validator node. When I do a Chain state > validatorSet > validators(): Option<Vec<AccountId>> I see:
```
[
  5DRZaKoY8hufAGN2yj4nfbGeVA2k7UTXNN9LFkdPYifkstyx,
  5HU576QAo9urobUfF5n1vNQMYfV1rrcD9uytf3eVvQgJFtTf,
  5CkjuTwGyzyXoWAYFSxZVnVHKxjxYRDiTZvCw7fjDHN6vjmv,
  5CkjuTwGyzyXoWAYFSxZVnVHKxjxYRDiTZvCw7fjDHN6vjmv
]
```

**Comment by davxy** [comment]
The issue is mostly caused by the sum of the following:
- capability to add duplicates to the validator list (this can also cause unfairness in block authoring)
- the delete function contains a snip that iterates over a vector that is not the same as the one we're removing elements from (this is has been obviously done due to the fact that we can't delete stuff from the vector we're iterating on). 

Possible solutions:
1. break as soon as an entry is delete
2. forbid duplicates

The question is: are duplicates allowed "by design"? Or is this a bug?

**Comment by gautamdhameja** [comment]
I have added a check for duplicates, and previously made some changes to avoid repeated session rotation.
The fix is in the https://github.com/gautamdhameja/substrate-validator-set/tree/fix-duplicates branch.
Let me know if the segmentation fault is still reproducible with this branch. I am not able to reproduce it at my side.

**Comment by davxy** [comment]
Indeed the crash thing was already fixed by commit: d11f58bfa0424b6efe29dffa0bc0c387dcb1bdaf 

The fix has been introduced by replacing this:

```
            for (i, v) in validators.clone().into_iter().enumerate() {
                if v == validator_id {
                    validators.swap_remove(i);
                }
            }
```

With this:

```
            validators.retain(|v| *v != validator_id);   
```                                                                      
---

Anyway, the duplicated entries issue has been fixed as well by the last commit. Thank you.

**Comment by gautamdhameja** [comment]
@davxy Thanks for confirming. Closing this.
